core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -780,6 +786,22 @@ object IO {
     override def tag = Tags.Ensuring
   }
 
+  final class RunSync[E, A] private[IO] (val io: IO[E, A])
```

Comment:
It may be possible to generalise `Redeem` instead of creating a new primitive.

Reply:
After thinking about this for several days, I think @regiskuckaertz is correct.

Ultimately what we want is a generalized `redeem`:

```scala
final class Redeem[E1, E2, A, B] private[IO] (
  val value: IO[E1, A],
  val term: List[Throwable] => IO[E2, B],
  val fail: E1 => IO[E2, B],
  val succ: A => IO[E2, B])
```

where obviously the exact structure will depend on the outcome of #342, and which one gets merged first (this one or that one).

With such a generalization, expressing `run` becomes trivial, and it will be possible to create very fast versions of some operations that are currently slow, and there will be less disruption and fewer edge cases to worry about in `RTS`.